CPLB userspace reverse proxy load balancer

juanluisvaladas opened the pull request:
## Description

Implement userspace CPLB reverse proxy load balancer.

Due to the keepalived virtualservers using IPVS for the load balancing, we simply couldn't make it work on some environments and needed to implement a userspace reverse proxy.

Fixes #4700 

## Type of change

<!-- check the related options -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## How Has This Been Tested?

- [x] Manual test
- [x] Auto test added

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

## Checklist:

- [x] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [x] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

twz123 commented on docs/cplb.md:
````diff
@@ -359,3 +374,150 @@ worker-0.k0s.lab       Ready    <none>          8m51s   v{{{ extra.k8s_version }
+for instance if a cluster has an `172.17.0.102/16` address and the interface is `eth0`,
+the expected output is similar to:
+
+```shell
````
```suggestion
```console
```

Same for the other terminal examples below.

juanluisvaladas commented on pkg/component/controller/cplb/tcpproxy/tcpproxy_test.go:
```diff
@@ -0,0 +1,226 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Modifications made by Mirantis Inc., 2024.
```
The header is copied and pasted from the email thread from the legal team. It also looks weird to me too, but they said:

> You should follow what is stated in section 4 (Redistribution) of the Apache 2.0. license. 
> <snip>
> 2) For cases where you are copying files from the Google project and making modifications, you need to add a notice that modifications were made by Mirantis, and year of modifications (a comment at the beginning of the file), add Google notice and then Mirantis notice, and license information.

juanluisvaladas commented on inttest/cplb-userspace-extaddr/cplbuserspaceextaddr_test.go:
```diff
@@ -0,0 +1,225 @@
+		s.PutFile(s.ControllerNode(idx), "/tmp/k0s.yaml", fmt.Sprintf(haControllerConfig, lb, lb))
+
+		// Note that the token is intentionally empty for the first controller
+		s.Require().NoError(s.InitController(idx, "--config=/tmp/k0s.yaml", "--enable-worker", "--disable-components=metrics-server,endpoint-reconciler", joinToken))
```
I removed the flag. I kept it because it was copy and paste from some other test.

juanluisvaladas commented on inttest/cplb-userspace/cplbuserspace_test.go:
```diff
@@ -0,0 +1,223 @@
+	}
+
+	// Create a worker join token
+	workerJoinToken, err := s.GetJoinToken("worker")
```
Modified

juanluisvaladas commented on inttest/cplb-userspace/cplbuserspace_test.go:
```diff
@@ -0,0 +1,223 @@
+
+	// Verify that controller+worker nodes are working normally.
+	s.T().Log("waiting to see CNI pods ready")
+	s.Require().NoError(common.WaitForKubeRouterReady(s.Context(), client), "kube router did not start")
```
Fixed, I copied and pasted a bit too many things from other tests...

juanluisvaladas commented on inttest/cplb-userspace/cplbuserspace_test.go:
```diff
@@ -0,0 +1,223 @@
+	s.Require().Len(signatures, 3, "Expected 3 different signatures, got %d", len(signatures))
+}
+
+// getLBAddress returns the IP address of the controller 0 and it adds 100 to
+// the last octet unless it's bigger or equal to 154, in which case it
+// subtracts 100. Theoretically this could result in an invalid IP address.
```
Modified it for:

```
// getLBAddress returns the IP address of the controller 0 and it adds 100 to
// the last octet unless it's bigger or equal to 154, in which case it
// subtracts 100. Theoretically this could result in an invalid IP address.
// This is so that get a virtual IP in the same subnet as the controller.

```

juanluisvaladas commented on inttest/cplb-userspace/cplbuserspace_test.go:
```diff
@@ -0,0 +1,223 @@
+// subtracts 100. Theoretically this could result in an invalid IP address.
+func (s *CPLBUserSpaceSuite) getLBAddress() string {
+	ip := s.GetIPAddress(s.ControllerNode(0))
+	parts := strings.Split(ip, ".")
```
Good point, done

juanluisvaladas commented on pkg/apis/k0s/v1beta1/cplb.go:
```diff
@@ -67,6 +67,14 @@ type KeepalivedSpec struct {
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	// +optional
+	UserSpaceProxyPort int `json:"userSpaceProxyBindPort,omitempty"`
```
That's a good point, the PR is already huge and there is another small PR waiting for this one to be merged for unicast VRRP and probably other 2 for known issues blocked on this one.

How about I add that in the subsequent PR?

juanluisvaladas commented on pkg/component/controller/cplb/tcpproxy/tcpproxy.go:
```diff
@@ -0,0 +1,492 @@
+
+	// ListenFunc optionally specifies an alternate listen
+	// function. If nil, net.Dial is used.
+	// The provided net is always "tcp".
```
This is the interface for `net.Listen`, we could remove the field but we need for the tests. I'd rather leave this as it is.

jnummelin replied:
> This is the interface for net.Listen

Would be more clear if there's a mention that in the comment :)

juanluisvaladas replied:
Good point. Added

juanluisvaladas commented on pkg/component/controller/cplb/tcpproxy/tcpproxy.go:
```diff
@@ -0,0 +1,492 @@
+// The order that routes are added in matters; each is matched in the order
+// registered.
+type Proxy struct {
+	configs map[string]*config // ip:port => config
```
Since it's a private field, how about we leave the map as it is (a string is more convinient than a struct for the map) and we use netip.AddrPort in the public functions?

juanluisvaladas commented on pkg/component/controller/cplb/tcpproxy/tcpproxy.go:
```diff
@@ -0,0 +1,492 @@
+// It returns whether it matched purely for testing.
+func (p *Proxy) serveConn(c net.Conn) bool {
+	br := bufio.NewReader(c)
+	for route := range p.routesChan {
```
I didn't do that because initially because SetRoutes and AddRoutes were not thread safe. Since we're making them threadsafe now I'll do it because it does make things simpler.

Since we have to lock to. read `cfg.Routes`anyway, instead of using an `atomic.Uint32` I just used an atomic int which should get the

juanluisvaladas replied:
I didn't do that because initially because SetRoutes and AddRoutes were not thread safe. Since we're making them threadsafe now I'll do it because it does make things simpler.

Since we have to lock to. read `cfg.Routes`anyway, instead of using an `atomic.Uint32` I just used int32 because it's locked anyway.

juanluisvaladas commented on pkg/component/controller/cplb/cplb_linux.go:
```diff
@@ -202,7 +225,7 @@ func (k *Keepalived) ensureDummyInterface(linkName string) error {
 		// it not existing. If we don't know what failed log it and attempt to
 		// create the link anyway.
-		if !errors.As(err, &netlink.LinkNotFoundError{}) {
+		if !errors.Is(err, &netlink.LinkNotFoundError{}) {
```
Good catch! reverting this.

juanluisvaladas commented on pkg/component/controller/cplb/cplb_linux.go:
```diff
@@ -321,7 +344,68 @@ func (k *Keepalived) generateKeepalivedTemplate() error {
+			}
+
+			cmd := exec.Command("/var/lib/k0s/bin/iptables", cmdArgs...)
+			if err := cmd.Run(); err != nil {
```
Good catch!

juanluisvaladas commented on pkg/component/controller/cplb/cplb_linux.go:
```diff
@@ -321,7 +344,68 @@ func (k *Keepalived) generateKeepalivedTemplate() error {
+	// forward everything to localhost
+	k.proxy.AddRoute(fmt.Sprintf(":%d", k.Config.UserSpaceProxyPort), tcpproxy.To(fmt.Sprintf("127.0.0.1:%d", k.APIPort)))
+
+	if err := k.proxy.Start(); err != nil {
```
Good catch, added it to `keepalived.Stop()`

juanluisvaladas commented on pkg/component/controller/cplb/cplb_linux.go:
```diff
@@ -321,7 +344,68 @@ func (k *Keepalived) generateKeepalivedTemplate() error {
+		for _, vipCIDR := range vrrp.VirtualIPs {
+			vip := strings.Split(vipCIDR, "/")[0]
+
+			cmdArgs := []string{
+				"-t", "nat", op, "PREROUTING", "-p", "tcp",
+				"-d", vip, "--dport", strconv.Itoa(k.APIPort),
+				"-j", "REDIRECT", "--to-port", strconv.Itoa(k.Config.UserSpaceProxyPort),
+			}
```
The clean up is done`Keepalived.Stop()` method.

Regarding dynamicConfig, cplb isn't managed with dynamiConfig and it's stated in the docs. At least until we have config v2 there is no way for us to make it compatible because the cplb configuration might differ per control plane node.

juanluisvaladas commented on pkg/component/controller/cplb/cplb_linux.go:
```diff
@@ -321,7 +344,68 @@ func (k *Keepalived) generateKeepalivedTemplate() error {
+	}
+
+	fmt.Println("Waiting for updateCh")
+	<-k.updateCh
```
`k.updateCh` is closed when the `Stop` method is called. Do we still need it?

jnummelin replied:
It might be bit more clear if the func gets the context as param, that way the stop case is bit more clear IMO

juanluisvaladas replied:
Fair enough. Added it.

juanluisvaladas commented:
I addressed most of the comments. As for the comments:

1-[Add tcpproxy](https://github.com/k0sproject/k0s/pull/5279/commits/a0fcb7e7c5d68a281119de96f6987df0fd816c99) remains unchanged for easier reviews
2- [Implement CPLB userspace reverse proxy LB](https://github.com/k0sproject/k0s/pull/5279/commits/3b3a1757c3d5f83496e290ccb246534b0afc2015) addresses all its comments. That includes adding `proxy.Stop()` to `keepalived.Stop()` reverting the changes of `errors.As` to `errors.Is`, clearer error messages, fixing the iptables path to use `k0sVars.BinDir`, and all the comments regarding the test
3- [Move the check for CPLB and externalAddress to cmd](https://github.com/k0sproject/k0s/pull/5279/commits/235bccd797aadb0ba63a4909e855a0ef72c3fe70) unifies the tests in one and only changes one error message to make it clearer, everything else remains unchanged (there were no more concerns.
4- [Overhaul CPLB docs](https://github.com/k0sproject/k0s/pull/5279/commits/7b3df0738d1a5b70a254c76c68bfa0a767a64e97) only changes the code blocks from `shell` to `console`
5- [Multiple refactors on cplb.tcpproxy](https://github.com/k0sproject/k0s/pull/5279/commits/ba2de2c13734f261bdd0986b98d10898b68ed5f7) addresses most of the concerns regarding the tcpproxy

There are a few things unchanged:
1- Copyright: I know it looks weird but it's literally copied and pasted from an email thread from the legal team. They said this is the right way of doing it 🤷 
2- Allowing to disable the userspace proxy: Because this is already far too large we need at least two subsequent PRs, one for more feature (unicast VRRP) and another one to fix a bug (etcd getting the VIP which has a conflict with the current code) how about we get that in one of the next PRs?
3- Having the tcpproxy binary as its own process: I'm not opposed, but I don't think it's a priority and doing it this way is simpler in terms of permissions (we need to query the API). If we think it's worth blocking then I'll take it out.
4-I didn't address this because I don't think it's required, but I'm not too sure: https://github.com/k0sproject/k0s/pull/5279/files#r1888710381

juanluisvaladas commented on cmd/controller/controller.go:
```diff
@@ -238,11 +239,23 @@ func (c *command) start(ctx context.Context) error {
 		}
 
+		if enableK0sEndpointReconciler {
+			return errors.New("control plane load balancing requires the component 'endpoint-reconciler' to be disabled")
```
OK, done

juanluisvaladas commented:
All issues addressed. Ready to review again.

jnummelin commented on pkg/component/controller/cplb/cplb_linux.go:
```diff
@@ -321,7 +344,68 @@ func (k *Keepalived) generateKeepalivedTemplate() error {
+	// forward everything to localhost
+	k.proxy.AddRoute(fmt.Sprintf(":%d", k.Config.UserSpaceProxyPort), tcpproxy.To(fmt.Sprintf("127.0.0.1:%d", k.APIPort)))
+
+	if err := k.proxy.Start(); err != nil {
```
That bin would access k8s api and need all the k8s client stuff built in, yet another time. So IMO it'll create bit unnecessary bloat without big benefits

jnummelin reviewed (approved):
LGTM at this stage. This is getting really large already so I think we should merge this at this stage and address any further things in subsequent PRs.

@juanluisvaladas let's make sure we have tickets for the remaining issues to be addressed
